velox/docs/develop/scalar-functions.rst

```diff
@@ -320,17 +320,68 @@ as much as possible to allow for faster compilation in codegen.
-so. These functions are inefficient because they require each value to be
-converted from a compact columnar representation into an STL container
-(std::vector, std::map, std::tuple). Also, it is not possible to define
-functions that accept generic arrays, maps or structs (e.g. map_keys,
-map_values, array_distinct, array_sort) as registerFunction requires all the
-signatures to be enumerated explicitly.
-
-Given this recommendation, we will not discuss how to define a simple function
-with complex type inputs or results.
+Complex types as inputs:
+************************
+Input complex types are represented in the simple function interface using light-weight lazy
+access abstractions that enable efficient direct access to the underlying data in Velox
+vectors.
+As mentioned earlier, the helper alias arg_type can be used in the function signature to
+map Velox types to the corresponding input types. The table below shows the actual types that are
+used to represent inputs of different complex types.
+
+==========  ==============================  =============================
+Velox Type  C++ Argument Type               C++ Actual Argument Type
+==========  ==============================  =============================
+ARRAY       arg_type<Array<E>>              ArrayView<VectorOptionalValueAccessor<VectorReader<E>>>>
+MAP         arg_type<Map<K,V>>              MapView<arg_type<K>, VectorOptionalValueAccessor<VectorReader<V>>>
+ROW         arg_type<Row<T1, T2, T3,...>>   RowView<VectorOptionalValueAccessor<arg_type<T1>>...>>
+==========  ==============================  =============================
+
+The view types are designed to have interfaces similar to those of std::containers, in fact in most cases
+they can be used as a drop in replacement. The table bellows shows the mapping between the Velox type and
+the corresponding similar std type.
+
+For example: a *Map<Row<int, int>, Array<float>>* is similar to const
+*std::map<std:::tuple<int, int>, std::vector<float>>*.
+
+===========      ======================================
+Lazy Input       Type  std Similar Object
+===========      ======================================
+ArrayView        const std::vector<std::optional<V>>
+MapView          const std::map<K, std::optional<V>>
+RowView          const std::tuple<std::optional<T1>...>
+===========      ======================================
+
+
+
+**1- VectorOptionalValueAccessor<VectorReader<E>>**:
+
+VectorOptionalValueAccessor is an *std::optional* like object that provides that lazy access to the nullity and
+value of the underlying Velox vector at a specific index. Currently, it is used to represent elements of input arrays
+and values in the input maps. Note that keys in the map are assumed to be not nullable in Velox.
+
+The object supports the following methods:
+- arg_type<E> value()      : unchecked access to the underlying value.
+- arg_type<E> operator *() : unchecked access to the underlying value.
```

Comment:
You might want to point out that there are subtle differences between this and and actual *, for e.g if I have a 2d input array, then something like ** wont work.